pkg/cover/canonicalizer.go

```diff
@@ -152,26 +175,64 @@ func findModule(pc uint32, moduleKeys []uint32) (moduleIdx int) {
 	return moduleIdx - 1
 }
 
-func (convert *Convert) convertPCs(cov []uint32, sign signal.Serial) {
+func (convert *Convert) convertPCs(cov []uint32, sign signal.Serial) ([]uint32, signal.Serial) {
 	// Convert coverage.
-	for idx, pc := range cov {
-		cov[idx] = convert.convertPC(pc)
+	var retCov []uint32
+	convCtx := &convertContext{convert: convert}
+	for _, pc := range cov {
+		if newPC, ok := convCtx.convertPC(pc); ok {
+			retCov = append(retCov, newPC)
+		}
+	}
+	if msg := convCtx.discarded(); msg != "" {
+		log.Logf(4, "error in PC conversion: %v", msg)
 	}
 	// Convert signals.
+	retSign := &signal.Serial{}
+	convCtx = &convertContext{convert: convert}
 	for idx, elem := range sign.Elems {
-		sign.UpdateElem(idx, convert.convertPC(uint32(elem)))
+		if newSign, ok := convCtx.convertPC(uint32(elem)); ok {
+			retSign.AddElem(newSign, sign.Prios[idx])
+		}
 	}
+	if msg := convCtx.discarded(); msg != "" {
+		log.Logf(4, "error in signal conversion: %v", msg)
+	}
+	return retCov, *retSign
+}
+
+func (cc *convertContext) discarded() string {
+	if cc.errCount == 0 {
+		return ""
+	}
+	errMsg := fmt.Sprintf("discarded 0x%x (and %v other PCs) during conversion", cc.errPC, cc.errCount)
+	return fmt.Sprintf("%v; not found in module map", errMsg)
 }
 
-func (convert *Convert) convertPC(pc uint32) uint32 {
-	moduleIdx := findModule(pc, convert.moduleKeys)
+func (cc *convertContext) discard(pc uint32) {
+	cc.errCount += 1
+	if cc.errPC == 0 {
+		cc.errPC = pc
+	}
+}
+
+func (cc *convertContext) convertPC(pc uint32) (uint32, bool) {
```

Comment:
Nit: I'd still leave most of `convertPC` in `Convert` in order to avoid multiple `cc.convert` and multiple `discard()` calls. `convertContext.convertPC` would then just be:

```
pc, ok := cc.convert.convertPC(pc)
if !ok {
 cc.discard(pc)
}
return pc, ok
```